File: node/src/git/diff-parser.ts
import { execSync } from 'child_process';
import type { DiffStatsMap } from '../types.js';

/**
 * Runs `git log --numstat` and accumulates line-level churn per file.
 *
 * Returns: Map<filename, { additions: number, deletions: number }>
 */
export function parseDiff(
  cwd: string,
  since: string = '12 months ago',
  pathFilter: string | null = null
): DiffStatsMap {
  const sinceArg = since ? `--since="${since}"` : '';
  const pathArg = pathFilter ? `-- "${pathFilter}"` : '';

  let output: string;
  try {
    output = execSync(
      `git log --format="COMMIT|%H" --numstat ${sinceArg} ${pathArg}`,
      { cwd, encoding: 'utf8', maxBuffer: 200 * 1024 * 1024 }
    );
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    throw new Error(`git numstat failed: ${message}`);
  }

  return parseNumstatOutput(output);
}

function parseNumstatOutput(output: string): DiffStatsMap {
  const stats: DiffStatsMap = new Map();

  for (const line of output.split('\n')) {
    const trimmed = line.trim();

    // Skip blank lines and commit header lines
    if (!trimmed || trimmed.startsWith('COMMIT|')) continue;

    // numstat format: <additions>\t<deletions>\t<filename>
    const parts = trimmed.split('\t');
    if (parts.length < 3) continue;

    // Binary files are shown as '-'
    if (parts[0] === '-' || parts[1] === '-') continue;

    const additions = parseInt(parts[0] ?? '0', 10);
    const deletions = parseInt(parts[1] ?? '0', 10);
    if (isNaN(additions) || isNaN(deletions)) continue;

    const rawFilename = parts[2];
    if (!rawFilename) continue;

    const filename = normalizeNumstatFilename(rawFilename);
    if (!filename) continue;

    const existing = stats.get(filename) ?? { additions: 0, deletions: 0 };
    stats.set(filename, {
      additions: existing.additions + additions,
      deletions: existing.deletions + deletions,
    });
  }

  return stats;
}

function normalizeNumstatFilename(raw: string): string | null {
  // Handle "src/{old => new}/file.js"
  if (raw.includes('{') && raw.includes('=>')) {
    const normalized = raw
      .replace(/\{[^}]+ => ([^}]+)\}/, '$1')
      .replace(/\/\//g, '/');
    return normalized.includes('{') ? null : normalized.trim();
  }

  // Handle "old-name => new-name"
  if (raw.includes(' => ')) {
    return raw.split(' => ').pop()?.trim() ?? null;
  }

  return raw.trim() || null;
}

File: rust/src/git/diff_parser.rs
use std::path::Path;
use std::process::Command;
use crate::types::DiffStatsMap;

/// Runs `git log --numstat` and accumulates line-level churn per file.
pub fn parse_diff(
    cwd: &Path,
    since: &str,
    path_filter: Option<&str>,
) -> Result<DiffStatsMap, String> {
    let mut args: Vec<String> = vec![
        "log".into(),
        "--format=COMMIT|%H".into(),
        "--numstat".into(),
    ];

    if !since.is_empty() {
        args.push(format!("--since={since}"));
    }

    if let Some(p) = path_filter {
        args.push("--".into());
        args.push(p.into());
    }

    let output = Command::new("git")
        .args(&args)
        .current_dir(cwd)
        .output()
        .map_err(|e| format!("Failed to run git numstat: {e}"))?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        return Err(format!("git numstat failed: {stderr}"));
    }

    let text = String::from_utf8_lossy(&output.stdout);
    Ok(parse_numstat_output(&text))
}

fn parse_numstat_output(output: &str) -> DiffStatsMap {
    let mut stats: DiffStatsMap = DiffStatsMap::new();

    for line in output.lines() {
        let trimmed = line.trim();
        if trimmed.is_empty() || trimmed.starts_with("COMMIT|") {
            continue;
        }

        // numstat format: <additions>\t<deletions>\t<filename>
        let parts: Vec<&str> = trimmed.splitn(3, '\t').collect();
        if parts.len() < 3 { continue; }

        // Binary files show '-'
        if parts[0] == "-" || parts[1] == "-" { continue; }

        let additions: usize = parts[0].parse().unwrap_or(0);
        let deletions: usize = parts[1].parse().unwrap_or(0);

        let filename = normalize_numstat_filename(parts[2]);
        if let Some(fname) = filename {
            let entry = stats.entry(fname).or_default();
            entry.additions += additions;
            entry.deletions += deletions;
        }
    }

    stats
}

fn normalize_numstat_filename(raw: &str) -> Option<String> {
    if raw.contains('{') && raw.contains("=>") {
        let re = once_cell::sync::Lazy::force(&RENAME_RE);
        let result = re.replace(raw, "$1").replace("//", "/");
        return if result.contains('{') { None } else { Some(result.trim().to_string()) };
    }
    if raw.contains(" => ") {
        return raw.split(" => ").last().map(|s| s.trim().to_string());
    }
    let t = raw.trim();
    if t.is_empty() { None } else { Some(t.to_string()) }
}

static RENAME_RE: once_cell::sync::Lazy<regex::Regex> =
    once_cell::sync::Lazy::new(|| regex::Regex::new(r"\{[^}]+ => ([^}]+)\}").unwrap());
